File: buildTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        
      int size = inorder.size();
        return buildTree_helper(inorder,postorder,0,size-1,0,size-1);
        
    }
    TreeNode* buildTree_helper(vector<int>& inorder, vector<int>& postorder, int inS,int inE, int posS,int posE)
    {
        if(inS>inE)
            return NULL;
        
        int rootData = postorder[posE];
        
        int lposS = posS;
        int rootIndex = -1;
        int i;
        
        for(i = inS;i<=inE;i++)
        {
            if(inorder[i]==rootData)
            {
                rootIndex = i;
                break;
                
            }
        }
        
        int linS = inS;
        int linE = rootIndex-1;
        int leftSubTreeLength = linE - linS + 1;
        int lposE = lposS+leftSubTreeLength-1; 
        int rinS = rootIndex+1;
        int rinE  =  inE;
        int rposS  =  linE+1;
        int rposE  =  posE-1;
        
        TreeNode* root = new TreeNode(rootData);
        
        root->left = buildTree_helper(inorder,postorder,linS,linE,lposS,lposE);
        
        root->right = buildTree_helper(inorder,postorder,rinS,rinE,rposS,rposE);
        
        return root;
            
            
    }
};

File: buildTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
    
        # preorder (Root left right)
        # inorder (left root right)
        
        preOrderLen = len(preorder)
        inOrderLen = len(inorder)
        
        if preOrderLen!=inOrderLen or inOrderLen == None or preOrderLen == None or preOrderLen==0 or inOrderLen==0 :
            return None
        
        root = TreeNode(preorder[0])
        rootData = root.val
        
        rootIndex = inorder.index(rootData)
        
        root.left = self.buildTree(preorder[1:rootIndex+1],inorder[:rootIndex])
        root.right = self.buildTree(preorder[rootIndex+1:],inorder[rootIndex+1:])
        
        return root
        
        
        
    
